src/tls.cpp

```diff
@@ -255,35 +264,60 @@ void tls_exit(void) {
     }
 }
 
-void *tls_connect(int fd) {
+int tls_connect_or_accept(SSL* ssl) {
+    int ret = 0;
+    bool try_again = false;
+    auto ssl_func = (SSL_is_server(ssl) ? SSL_accept : SSL_connect);
+    
+    do {
+        try_again = false;
+        ret = ssl_func(ssl);
+        if (ret < 0 && !g_b_exit) {
+            int err = SSL_get_error(ssl, ret);
+            if (IS_TLS_ERR_WANT_RW(err) || err == SSL_ERROR_WANT_CONNECT) {
+                int which =
+                    ((err == SSL_ERROR_WANT_READ || err == SSL_ERROR_WANT_CONNECT) ? 1 : 0) |
+                    ((err == SSL_ERROR_WANT_WRITE || err == SSL_ERROR_WANT_CONNECT) ? 2 : 0);
+               
+                log_dbg("tls_connect_or_accept waiting for socket");
+                int rc = wait_for_single_socket(SSL_get_fd(ssl), which);
+                if (0 > rc && errno != EINTR) {
+                    log_err("Failed to wait for event while establishing TLS");
+                } else if (!g_b_exit) {
+                    try_again = true;
+                }
+            } else {
+                tls_log_err("ERROR: Failed to establish TLS (%d,%d)", err, errno);
+            }
+        }
+    } while (try_again);
+
+    return ret;
+}
+
+void* tls_establish(int fd) {
     SSL *ssl = NULL;
 
     if (!ssl_ctx) {
+        log_msg("ERROR: Failed tls_establish(), no ssl_ctx");
         goto err;
     }
 
     ssl = SSL_new(ssl_ctx);
     if (!ssl) {
-        log_err("Failed SSL_new()");
+        log_msg("ERROR: Failed SSL_new()");
```

Comment:
Why not to use `log_err()`

Reply:
Because log_err prints errno and in this case it always will have a misleading information (some older socket call or etc).
The failure of SSL_new is not related to errno in this case.

Reply:
Will use log_err